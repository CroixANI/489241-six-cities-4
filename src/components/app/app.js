import React from "react";
import Main from "../main/main.js";
import PropTypes from "prop-types";

const App = (props) => {
  const {titles, onOfferTitleClick} = props;

  return (
    <Main titles={titles} onOfferTitleClick={onOfferTitleClick} />
  );
};

App.propTypes = {
  titles: PropTypes.arrayOf(
      PropTypes.string.isRequired
  ).isRequired,
  onOfferTitleClick: PropTypes.func.isRequired,
  offers: PropTypes.arrayOf(
      PropTypes.shape({
        title: PropTypes.string.isRequired,
        price: PropTypes.number.isRequired,
        rating: PropTypes.number.isRequired,
        city: PropTypes.string.isRequired,
        apartmentType: PropTypes.string.isRequired,
        offerTag: PropTypes.string.isRequired,
        isFavorite: PropTypes.bool.isRequired,
      })
  ),
};

export default App;
