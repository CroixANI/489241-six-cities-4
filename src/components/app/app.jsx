import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {Switch, Route, BrowserRouter} from "react-router-dom";
import {connect} from "react-redux";

import Main from '../main/main.jsx';
import Offer from '../offer/offer.jsx';
import {ActionCreator} from "../../reducer.js";

class App extends PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/">
            {this._renderApp()}
          </Route>
        </Switch>
      </BrowserRouter>
    );
  }

  _renderApp() {
    const {currentOfferId, offers, cities, selectedCity, onCityClick, onOfferClick} = this.props;
    const foundOffer = offers.find((offer) => offer.id === currentOfferId);

    if (foundOffer) {
      return <Offer offer={foundOffer} onOfferTitleClick={onOfferClick} />;
    } else {
      return <Main offers={offers} cities={cities} selectedCity={selectedCity} onCityClick={onCityClick} onOfferTitleClick={onOfferClick} />;
    }
  }
}

App.propTypes = {
  onCityClick: PropTypes.func.isRequired,
  onOfferClick: PropTypes.func.isRequired,
  cities: PropTypes.arrayOf(PropTypes.string).isRequired,
  selectedCity: PropTypes.string.isRequired,
  currentOfferId: PropTypes.number,
  offers: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        title: PropTypes.string.isRequired,
        price: PropTypes.number.isRequired,
        rating: PropTypes.number.isRequired,
        type: PropTypes.string.isRequired,
        luxuryType: PropTypes.string.isRequired,
        isBookmarked: PropTypes.bool.isRequired,
        images: PropTypes.arrayOf(PropTypes.string).isRequired,
        location: PropTypes.shape({
          city: PropTypes.string.isRequired,
          latitude: PropTypes.number.isRequired,
          longitude: PropTypes.number.isRequired
        }).isRequired,
      })
  ),
};

const mapStateToProps = (state) => ({
  offers: state.filteredOffers,
  cities: state.cities,
  selectedCity: state.city,
  currentOfferId: state.currentOfferId
});

const mapDispatchToProps = (dispatch) => ({
  onCityClick(city) {
    dispatch(ActionCreator.changeCity(city));
    dispatch(ActionCreator.listOffers());
  },
  onOfferClick(offerId) {
    dispatch(ActionCreator.changeCurrentOffer(offerId));
  }
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
