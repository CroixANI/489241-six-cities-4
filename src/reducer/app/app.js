import {SORT_TYPE} from '../../data/constants';

const initialState = {
  city: ``,
  currentOfferId: null,
  sortType: SORT_TYPE.POPULAR,
  errorMessage: null,
};

const ActionType = {
  CHANGE_CITY: `CHANGE_CITY`,
  CHANGE_SORT_TYPE: `CHANGE_SORT_TYPE`,
  CHANGE_CURRENT_OFFER: `CHANGE_CURRENT_OFFER`,
  SET_ERROR_MESSAGE: `SET_ERROR_MESSAGE`
};

const ActionCreator = {
  changeCity: (city) => ({
    type: ActionType.CHANGE_CITY,
    payload: city
  }),
  changeSortType: (sortType) => ({
    type: ActionType.CHANGE_SORT_TYPE,
    payload: sortType
  }),
  changeCurrentOffer: (offerId) => ({
    type: ActionType.CHANGE_CURRENT_OFFER,
    payload: offerId
  }),
  setErrorMessage: (errorMessage) => ({
    type: ActionType.SET_ERROR_MESSAGE,
    payload: errorMessage
  })
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_SORT_TYPE:
      return Object.assign({}, state, {
        sortType: action.payload
      });
    case ActionType.CHANGE_CITY:
      return Object.assign({}, state, {
        city: action.payload
      });
    case ActionType.CHANGE_CURRENT_OFFER:
      return Object.assign({}, state, {
        currentOfferId: action.payload
      });
    case ActionType.SET_ERROR_MESSAGE:
      return Object.assign({}, state, {
        errorMessage: action.payload
      });
  }

  return state;
};

export {reducer, ActionType, ActionCreator};
