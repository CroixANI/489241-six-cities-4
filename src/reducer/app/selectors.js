import {createSelector} from 'reselect';

import {SORT_TYPE} from '../../data/constants';
import {getOffers, getCities} from '../data/selectors';
import NameSpace from '../name-space';

const NAME_SPACE = NameSpace.APP;

export const getSortType = (state) =>
  state[NAME_SPACE].sortType || SORT_TYPE.POPULAR;

export const getCity = (state) => {
  const city = state[NAME_SPACE].city;
  if (city) {
    return city;
  }

  const cities = getCities(state);
  if (cities && cities.length > 0) {
    return cities[0].name;
  }

  return ``;
};

export const getCurrentOfferId = (state) =>
  state[NAME_SPACE].currentOfferId;

export const getCurrentOffer = createSelector(
    getCurrentOfferId,
    getOffers,
    (offerId, offers) => offers.find((offer) => offer.id === offerId));

export const getFilteredOffers = createSelector(
    getCity,
    getOffers,
    (currentCity, offers) => offers.filter((offer) => offer.location.city.name === currentCity));

const sortOffers = (offers, sortType) => {
  const copyOffers = [...offers];

  switch (sortType) {
    case SORT_TYPE.PRICE_HIGH_TO_LOW:
      return copyOffers.sort((a, b) => b.price - a.price);
    case SORT_TYPE.PRICE_LOW_TO_HIGH:
      return copyOffers.sort((a, b) => a.price - b.price);
    case SORT_TYPE.TOP_RATED:
      return copyOffers.sort((a, b) => b.rating - a.rating);
  }

  return offers;
};

export const getSortedOffers = createSelector(
    getSortType,
    getFilteredOffers,
    (sortType, filteredOffers) => sortOffers(filteredOffers, sortType));

export const getErrorMessage = (state) =>
  state[NAME_SPACE].errorMessage;
