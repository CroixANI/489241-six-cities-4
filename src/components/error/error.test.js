import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';

import Error from './error.jsx';
import NameSpace from '../../reducer/name-space';
import {ActionCreator} from '../../reducer/app/app';

const mockStore = configureStore([]);
const store = mockStore({
  [NameSpace.APP]: {
    errorMessage: null,
  },
});

describe(`Render Error component`, () => {
  it(`Render error component without any error`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <Error />
          </Provider>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Render error component with error message`, () => {
    store.dispatch(ActionCreator.setErrorMessage(`My message`));
    const tree = renderer
      .create(
          <Provider store={store}>
            <Error />
          </Provider>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

