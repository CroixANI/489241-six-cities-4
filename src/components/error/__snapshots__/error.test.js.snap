// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Error component Render error component with error message 1`] = `null`;

exports[`Render Error component Render error component without any error 1`] = `null`;
